File: 001-sorting/heap-sort/heap-sort.go
package main

import (
	"fmt"
)

var heapSize int

func main() {
	a := []int{10, 9, 2, 11, 1, 6}
	fmt.Println("Before sorting", a)
	heapSort(a)
	fmt.Println("After sorting", a)
}

func heapSort(a []int) {
	buildMaxHeap(a)
	for i := len(a) - 1; i >= 1; i-- {
		tmp := a[0]
		a[0] = a[i]
		a[i] = tmp
		heapSize = heapSize - 1
		maxHeapify(a, 0)
	}
}

func buildMaxHeap(a []int) {
	heapSize = len(a)
	mid := (len(a) - 1) / 2
	for i := mid; i >= 0; i-- {
		maxHeapify(a, i)
	}
}

func maxHeapify(a []int, i int) {
	l := left(i)
	r := right(i)
	largest := i
	if l < heapSize && a[l] > a[largest] {
		largest = l
	}

	if r < heapSize && a[r] > a[largest] {
		largest = r
	}

	if largest != i {
		tmp := a[i]
		a[i] = a[largest]
		a[largest] = tmp
		maxHeapify(a, largest)
	}

}

func left(i int) int {
	return 2*i + 1
}

func right(i int) int {
	return 2*i + 2
}

File: 001-sorting/heap-sort/heap-sort.js
var heapSize = 0;

function heapSort(a) {
  buildMaxHeap(a);
  for (i = a.length - 1; i >= 1; i--) {
    var tmp = a[0];
    a[0] = a[i]
    a[i] = tmp;
    heapSize = heapSize - 1;
    maxHeapify(a, 0);
  }
}

function buildMaxHeap(a) {
  heapSize = a.length;
  var mid = Math.floor((a.length-1)/2);
  for (i = mid; i >= 0; i--) {
    maxHeapify(a, i);
  }
}

function maxHeapify(a, i) {
  var l = left(i);
  var r = right(i);
  var largest = i;
  if (l < heapSize && a[l] > a[largest]) {
    largest = l;
  }
  if (r < heapSize && a[r] > a[largest]) {
    largest = r;
  }
  if (largest != i) {
    var tmp = a[i];
    a[i] = a[largest];
    a[largest] = tmp;
    maxHeapify(a, largest);
  }
}

function left(i) {
  return 2*i + 1;
}

function right(i) {
  return 2*i + 2;
}

var a = [10, 23, 1, 5, 23, 67, 8 ,9, 11, -2, 234, 123];
console.log("Before sorting", a);
heapSort(a);
console.log("After sorting", a);